torch/csrc/jit/codegen/cuda/lower_predicate.cpp

```diff
@@ -324,17 +350,18 @@ bool PredicateElimination::needsPredicate(Expr* expr) const {
     return false;
   });
 
-  // Predicates Welford ops
-  filters.emplace_back([](Expr* expr) { return expr->isA<WelfordOp>(); });
-
   // If this is a reduction, and if we omit the predicate for the
   // input, the input may have a garbabe value, which must not be used
-  // for this reduction. However, if the input is also an output of
-  // another reduction with the same binary op, which is a common
-  // pattern with rfactor, the input should be safe to use with no
-  // predication.
+  // for this reduction. However, it is still legal to omit its
+  // predicate when: 1) the predicate of the input is not omitted and
+  // 2) the input can be initialized to the init value of this
+  // reduction. When the input is the output of another reduciton, the
+  // input is initialized to the init value of the reduction, so the
+  // two reductions must use the same init value.
+  // See FusionPredicateElimination3 and FusionPredicateElimination4
+  // for concrete examples.
```

Comment:
I'll refactor this function in a follow-up PR. I don't like to have this large function, although it's conceptually separated into a series of anonymous lambdas.